### PR TITLE
Added Logfile support

### LINKED ISSUE #274: Redirect log entries to file
GD2 should be able to log to a file

### PR DESCRIPTION
**Pull request by aravindavk**
New CLI flag "logfile" added. Default value is "-" which writes to
StdErr. Logfile can be specified using,

    ./glusterd2 --logfile glusterd.log

NOTE: Use the flag `--logdir` to configure the log directory if required

    ./glusterd2 --logfile glusterd.log --logdir /var/log/glusterfs

Also added support for logrotate, if log file rotated, notify the running
glusterd process by sending SIGHUP(`kill -SIGHUP $PID`)

Closes #274
Signed-off-by: Aravinda VK <avishwan@redhat.com>

### CODE REVIEW

**Comment by kshlm** [comment]
add to whitelist

**Comment by kshlm** [comment]
ok to test

File: main.go [hunk 1]
```diff
@@ -30,7 +31,20 @@ func main() {
...
+	logFileName, _ := flag.CommandLine.GetString("logfile")
+
+	if logFileName == "-" {
+		initLog(logLevel, os.Stderr)
+	} else {
+		logFilePath := path.Join(logdir, logFileName)
+		logFile, logFileErr := openLogFile(logFilePath)
+		if logFileErr != nil {
+			initLog(logLevel, os.Stderr)
+			log.WithError(logFileErr).Fatalf("Failed to open log file %s", logFilePath)
```
Comment (prashanthpai) [inline main.go]: I know that `Fatal` will internally log and do a `os.Exit(1)`. I think `Fatalf` does the same. If so, the program will exit. Please check.

File: logging.go [hunk 2]
```diff
@@ -2,16 +2,51 @@ package main
...
+		if logWriter != nil {
+			logWriter.Close()
+		}
+
+		// Reopen the log file again and update logWriter
+		logFilePath := path.Join(logdir, logFileName)
+		logFile, logFileErr := openLogFile(logFilePath)
+		if logFileErr != nil {
+			initLog(logLevel, os.Stderr)
+			log.WithError(logFileErr).Fatalf("Failed to open log file %s", logFilePath)
```
Comment (prashanthpai) [inline logging.go]: See comment below about using `Fatalf`

File: logging.go [hunk 3]
```diff
@@ -2,16 +2,51 @@ package main
...
+	"path"
 	"strings"
 
 	log "github.com/Sirupsen/logrus"
 )
 
-func initLog(logLevel string, out io.Writer) {
+var logWriter io.WriteCloser
+
+func reloadLog(logdir string, logFileName string, logLevel string) {
```
Comment (prashanthpai) [inline logging.go]: I think you can combine `reloadLog` and `initLog` in the same function `reloadLog`. I see a repetitive pattern in calls to `initLog(logLevel, os.Stderr)` after trying to open log file.

**Comment by kshlm** [comment]
test this pr

File: main.go (hunk 1 above)
Comment (aravindavk) [inline main.go]: Ok, what is expected here? glusterd should continue if it fails to open/reopen log file or it should log to Stderr and continue?

File: config.go [hunk 4]
```diff
@@ -36,6 +36,7 @@ func parseFlags() {
 	flag.String("localstatedir", "", "Directory to store local state information. (default: workdir)")
 	flag.String("rundir", "", "Directory to store runtime data. (default: workdir/run)")
 	flag.String("logdir", "", "Directory to store logs. (default: workdir/log)")
+	flag.String("logfile", "-", "Log file name. (default: -)")
```
Comment (aravindavk) [inline config.go]: Sure. Will make glusterd2.log as default. and accept `stderr` or `stdout` as inputs and use os.Stderr or os.Stdout respectively.

    --logfile=stderr
    --logfile=stdout
    --logfile=somelog.log

File: logging.go (hunk 3 above)
Comment (aravindavk) [inline logging.go]: Ok

File: main.go (hunk 1 above)
Comment (kshlm) [inline main.go]: Log to stderr and fail. This consistent with the rest of GlusterFS.

File: config.go (hunk 4 above)
Comment (prashanthpai) [inline config.go]: I meant to point to update the help text only. The rest is fine. YUpdate the help text to `Log file name. (logs to stderr by default)`.

File: main.go (hunk 1 above)
Comment (prashanthpai) [inline main.go]: If that's the case, then this is all good. Sorry for the confusion.

**Comment by prashanthpai** [comment]
@kshlm Feel free to merge this if you have no further review comments :+1:

**Pull request merged**